### PR TITLE
Minimal era for Run 3 PbPb and corresponding relVal workflows

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

Adds Run3_PbPb to run PbPb reconstruction with 2021 conditions.  This era contains the minimal modifications to produce a running workflow.  It uses the same modifiers as in the Run2_2018_pp_on_AA era.  Also adding 2021 heavy-ion MC relval workflows by copying the corresponding 2018 ones. 

#### PR validation:

Tested locally with the new relval workflow 159.1, after adding PR #29078 by hand.

### CODE REVIEW

**Comment by mandrenguyen** [comment]
@abaty

**Comment by boudoul** [comment]
Great , thanks @mandrenguyen ! Presumably we should define a workflow using it, do you want me to take care ?

**Comment by mandrenguyen** [comment]
@boudoul I can setup a relval workflow. Let's get some feedback first on this proposal for an era from @slava77 @perrotta

**Comment by boudoul** [comment]
Perfect, thank you

File: Configuration/Eras/python/Era_Run3_PbPb_cff.py
Comment (slava77) [inline Configuration/Eras/python/Era_Run3_PbPb_cff.py]: why not keep the same style as before and have "pp_on_AA" in the name[s]?

**Review by slava77 — COMMENTED** [review]
> except it removes the HI-specific track MVA calibrations, which do not yet exist in a 2021 global tag

why do they not exist for the 2021 GT?
IIRC, the regular tracking selectors are the same in 2021 as in 2018.
Did I miss some essential tracking changes that would invalidate HIMVASelector[s]?

File: Configuration/Eras/python/Era_Run3_PbPb_cff.py
Comment (mandrenguyen) [reply Configuration/Eras/python/Era_Run3_PbPb_cff.py]: pp_on_AA was intended to mean that we are running a pp-like reco on AA collisions.  This is in contrast to the dedicated heavy-ion reco, which was also still being maintained in 2018.  Now that such a dedicated reco sequence has been deprecated, I thought it best to simplify the naming.   I'm open to suggestion though.

**Comment by mandrenguyen** [comment]
> > except it removes the HI-specific track MVA calibrations, which do not yet exist in a 2021 global tag
> 
> why do they not exist for the 2021 GT?
> IIRC, the regular tracking selectors are the same in 2021 as in 2018.
> Did I miss some essential tracking changes that would invalidate HIMVASelector[s]?

In my tests I have used the standard pp GT (and corresponding beamspot) in my tests.   A crash in reco indicated to me that they do not appear to contain this record.   I assume there is not matching GT/beamspot set yet for heavy-ion beam conditions.   I don't believe there's anything fundamental.  One could simply add these records to the pp GTs as well.   Of course the selection  has to be retrained for Run 3 in any case.

**Comment by slava77** [comment]
> One could simply add these records to the pp GTs as well.

I prefer this solution; it seems more appropriate for the current state of the physics performance setup.

File: Configuration/Eras/python/Era_Run3_PbPb_cff.py
Comment (slava77) [reply Configuration/Eras/python/Era_Run3_PbPb_cff.py]: IIUC, the HeavyIon scenario still exists. So, there is still a clear case to still call this pp_on_AA. To me the naming should reflect the status of the release in a coherent way rather than a specific intention for the future.
I prefer a coherent migration instead of changing in a single point.

Also, would AA or HI still work, or does "PbPb" need to be used here?

**Comment by mandrenguyen** [comment]
@abaty Can you upload the HIMVA tags into the GT queue for 2021 realistic conditions, please?

File: Configuration/Eras/python/Era_Run3_PbPb_cff.py
Comment (mandrenguyen) [reply Configuration/Eras/python/Era_Run3_PbPb_cff.py]: I don't think there's much potential for confusion of --scenario HeavyIons with pbpb_run3, whereas most people will not understand the meaning of pp_on_AA, but I don't insist on the "pp_on_X" part.   AA denotes generic ion collisions.  Up to now we had PbPb and XeXe.  If we ran these today, we would like use the same pp_on_AA modifier on both, so I used AA to distinguish from pA.   However, in Run 3 we are likely to have Oxygen-Oxygen collisions, for which we would use the standard pp high lumi reconstruction settings.  So pp_on_AA doesn't make a whole lot of sense.   Should I call it pp_on_PbPb_run3?  Or we just live with pp_on_AA_2021, in the name of continuity?

**Comment by abaty** [comment]
@mandrenguyen I've asked Xiao Wang to upload the relevant tags to the GT queue.

**Comment by tocheng** [comment]
I prefer to have a separate GT for heavy Ion MC. AlCa can make a heavy Ion queue for 2021 based on pp and heavy Ion group can queue tags to add new conditions or replace conditions in pp GT.

**Comment by mandrenguyen** [comment]
@tocheng Is there a reason not to add this record to the pp GT?  We may want to test the HI era with various pp GTs in the future.  In any case, this is fine for us if @slava77  agrees.  What would be the ETA to produce a HI GT that is the same as the one pointed to by auto:phase1_2021_realistic, except that it has the add'l HIMVA records?

**Comment by tocheng** [comment]
@mandrenguyen What do you mean various pp GTs?
We could make a new queue  based on pp 2021 and you can add the missing tracker selector. The HI GT can be made soon after.

**Comment by mandrenguyen** [comment]
@tocheng I just meant that adding this record to pp GTs would allow us to run PbPb_run3 era with pp GTs without crashing.   That sounds like something that could come in handy to me.  I think we can drop this discussion though.    We will need the HI queue anyway, since there will eventually be beamspot tag to upload there.  Let's just get this era operational.   Once a HI GT is available I can update the autocond, and add relval wf.

**Comment by tocheng** [comment]
111X_mcRun3_2021_realistic_HI_Queue is ready and is equal to auto:phase1_2021
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/111X_mcRun3_2021_realistic_HI_Queue/110X_mcRun3_2021_realistic_v8

**Comment by Taburis** [comment]
Hi @tocheng, @mandrenguyen, and @slava77, the MVA tags added to the queue: 111X_mcRun3_2021_realistic_HI_Queue.

**Comment by christopheralanwest** [comment]
PR #29078 contains a new autoCond key, `auto:phase1_2021_realistic_hi`, with the requested global tag. A new relval workflow should be created that uses this new key.

**Comment by christopheralanwest** [comment]
assign pdmv

This PR should introduce a relval workflow that uses the `auto:phase1_2021_realistic_hi` GT from PR #29078 prior to merging these two PRs. We don't want to introduce new autoCond keys that are not tested by any relval workflow.

**Comment by mandrenguyen** [comment]
@christopheralanwest I can introduce a new relval workflow, but I'm not clear on your proposed order of operations.  If I add the workflow to this PR it won't pass tests until #29078 is merged right?  So shouldn't we merge #29078 and then update this PR (or even just make a 3rd PR with the relval wf)?

**Comment by chayanit** [comment]
> @christopheralanwest I can introduce a new relval workflow, but I'm not clear on your proposed order of operations. If I add the workflow to this PR it won't pass tests until #29078 is merged right? So shouldn't we merge #29078 and then update this PR (or even just make a 3rd PR with the relval wf)?

@mandrenguyen two PR can be tested together and no need to wait for one to be merged. I think you can add relval wf in this PR.

**Comment by mandrenguyen** [comment]
@christopheralanwest I have copied the 2018 HI wfs to the 2021 era and rolled back the use of the pp track MVA calibrations, since the HI ones are now available from the GT.   The driver options are as described in the original PR description.

**Comment by christopheralanwest** [comment]
please test workflow 159,159.1,159.2,159.3 with #29078

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen you may need to introduce the new baseDataSetRelease for 2021 string

**Review by chayanit — CHANGES_REQUESTED** [review]

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @chayanit  Yes, it would make sense to switch to 14 TeV.  Now that it occurs to me, it would be very useful to have one gen fragment that's run with both the standard pp settings and the HI era.   Would it be ok to add a TTbar_14TeV_TuneCP5 workflow with heavy-ion settings  for this purpose?   Regarding the other point, I will clean up the ZMM, which is no longer used.

Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen we do have these GEN fragments for 14TeV. Please update
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/QCD_Pt_80_120_14TeV_TuneCUETP8M1_cfi.py
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/PhotonJet_Pt_10_14TeV_TuneCUETP8M1_cfi.py
but not ZEEMM. Would you like to switch from ZEEMM to TTbar?

**Comment by boudoul** [comment]
Hello @mandrenguyen , just a small detail:  may you please update the title : In addition to 'Minimal Era for Run2 PbPb' you may want to add 'and corresponding RunTheMatrix Workflows' (or something like that) . Thanks !

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: Thanks!  Z bosons tend to be pretty useful for us, I'm not sure I would like get rid of these wfs.   I see that there are separate ZMM and ZEE gen fragments for 14 TeV already.  What if  switched to these and reduced the number of events for each wf by a factor of 2?  
While the 14 TeV fragments exist, are any of these workflows actually run in the IB?    That was my idea for adding ttbar.  On the other hand, the scope of this PR is already spreading beyond my initial intention, so maybe we can drop that idea.

Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen  Instead of split them to two more workflows, I would recommend to create new GEN fragment for ZEEMM 14TeV under Configuration/Generator/python. You can just copy this GEN fragment and update energy to 14TeV https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/ZEEMM_13TeV_TuneCUETP8M1_cfi.py
I don't think any of these workflows are included in the IB, usually we use TTbar. You can try to add and I will have a look. 

I have other comments elsewhere. Please also consider the changes. Thanks.

Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @chayanit Your other comments seem straightforward, so I will address them as soon as I can find some time :-)

File: Configuration/Eras/python/Era_Run3_PbPb_cff.py
Comment (slava77) [reply Configuration/Eras/python/Era_Run3_PbPb_cff.py]: I still think that ```pp_on_``` should still be a part of the era name. I'm fine to stay with PbPb if other options may need a different setup

File: Configuration/Eras/python/Modifier_pbpb_run3_cff.py
Comment (slava77) [inline Configuration/Eras/python/Modifier_pbpb_run3_cff.py]: this is not needed anymore.
If a placeholder is desired, a name following existing pattern would be pp_on_AA_run3 or pp_on_PbPb_run3 if finer specification is desired

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen your current setup will call baseDataSetRelease[9] string which is from Run2. I was asking you to implement new index of baseDataSetRelease but yes you're right we never produced any Run3 mixing yet and it will fail. You can continue with what you have for now.

**Comment by kpedro88** [comment]
please test

**Comment by christopheralanwest** [comment]
please abort

The tests should be run with PR #29078 and test the workflows introduced in this PR.

**Comment by christopheralanwest** [comment]
please test workflow 159,159.1,159.2,159.3,159.4,310,311,312 with #29078

**Comment by kpedro88** [comment]
@christopheralanwest in the future, please use the "test parameters" [bot command](http://cms-sw.github.io/cms-bot-cmssw-cmds.html) so that these special settings don't have to be included every time someone calls "please test"

**Comment by mandrenguyen** [comment]
@christopheralanwest 310 and related workflows fail b/c of the lack of an appropriate pileup_dataset.  The GEN-SIM output of the new workflow 159 will be used for this.  See the discussion with @chayanit in this PR's comments.

**Comment by christopheralanwest** [comment]
test parameters
workflow = 159,159.1,159.2,159.3,159.4
pull_request = #29078

**Comment by christopheralanwest** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/29032	5714287 
- changes in the reco category are trivial (an empty line added in RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py); the Era modifier name (not explicitly in reco category) is in line with the review comments.
- jenkins tests passed (where expected); there are no diffs to make in the results given that the rest of the PR is introducing a new workflow. The non-PU workflows ran OK already in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3e9604/5037/summary.html

**Comment by christopheralanwest** [comment]
please test

The global tag in #29078 needed to be updated to include an update to the HCAL zero suppression for Run 3.

**Comment by silviodonato** [comment]
+operations

**Comment by silviodonato** [comment]
Do you have any comments?
upgrade:  @kpedro88
pdmv:  @chayanit @pgunnell @zhenhu

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen I thought we agreed to have ZEEMM not separated. Could you add corresponding GEN fragment and update this part?

**Review by chayanit — CHANGES_REQUESTED** [review]

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @chayanit What is the logic of running them together?

Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen what is the logic to run them separate? We used to run them together in 2018 no?

Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @chayanit  The logic is to keep them in sync easily with the high lumi pp workflows.  I don't see why we should add an extra layer of complexity to anyone analyzing the output, as well as add additional gen fragments to maintain.  If you're worried about the extra processing, we can cut the number of events in half.

Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen ZEEMM has been used as HI workflow since 2018 and now you said it's adding extra layer of complexity by just selecting Z from muon or electron? If you want to keep them in sync with high lumi pp, you will need to update all workflows to TuneCP5.

Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: Yes, was the one who made the PR to add ZEEMM IIRC.  Can we not be stuck with it forever, please?

Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: So please remove ZEEMM line you commented out completely if you wanna move to separate workflows. How about the Tuning? Do we stay with the old tuning for HI?

Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: Ok, will do this afternoon, thanks!

Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @chayanit  I just pushed the change to CP5 for the ZEE and ZMM wfs (and removed the commented out wf).  I don't see CP5 fragments for the QCD and gamma-jet.  Did you want me to migrate those as well?

**Comment by slava77** [comment]
@cmsbuild please test

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (chayanit) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: @mandrenguyen it would be nice to have everything in the same setup. Thanks!

File: RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py
Comment (kpedro88) [inline RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py]: remove errant newline so this PR doesn't change this file for no reason

Comment (mandrenguyen) [reply RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py]: The offending newline is removed in the latest commit

File: Configuration/PyReleaseValidation/python/relval_steps.py
Comment (mandrenguyen) [reply Configuration/PyReleaseValidation/python/relval_steps.py]: I have added the CP5 fragments and updated the relevant workflows

**Comment by slava77** [comment]
@cmsbuild please test

notably, a reco signature is not needed anymore

**Comment by chayanit** [comment]
+1

**Comment by silviodonato** [comment]
+operations

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
Kind reminder for generators:  @alberto-sanchez @agrohsje @efeyazgan @mkirsano @qliphy @SiewYan

**Comment by qliphy** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**
